src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,887 @@
+            if (_spiDevice == null)
+            {
+                _spiDevice = new UnixSpiDevice(settings);
+            }
+            GoPiGo3Info = new GoPiGoInfo();
+            if (autoDetect == true)
+                try
+                {
+                    GoPiGo3Info.Manufacturer = GetManufacturer();
+                    GoPiGo3Info.Board = GetBoard();
+                    GoPiGo3Info.SoftwareVersion = GetFirmwareVersion();
+                    GoPiGo3Info.HardwareVersion = GetHardwareVersion();
+                    GoPiGo3Info.Id = GetIdHex();
+                }
+                catch (IOException ex)
+                {
+                    throw new IOException($"No SPI response. GoPiGo3 with address {SpiAddress} not connected.", ex);
+                }
+            if ((GoPiGo3Info.Manufacturer != ManufacturerName) || (GoPiGo3Info.Board != BoardName))
+                throw new IOException($"GoPiGo3 with address {SpiAddress} not connected.");
+            MotorGearRatio = DefaultMotorGearRatio;
+            EncoderTicksPerRotation = DefaultEncoderTicksPerRotation;
+            // Initialise the 2 Groove sensors
+            GrooveSensor = new GrooveSensor[2]
+            {
+                new GrooveSensor(GroovePort.Groove1),
+                new GrooveSensor(GroovePort.Groove2)
+            };
+        }
+
+        /// <summary>
+        /// Reset everything
+        /// </summary>
+        public void Dispose()
+        {
+            // Reset groove sensors
+            SetGrooveType(GroovePort.Both, GrooveSensorType.Custom);
+            SetGrooveMode(GroovePort.Both, GrooveInputOutput.InputDigital);
+            // Turn off the motors
+            SetMotorPower(MotorPort.Both, (byte)MotorSpeed.Float);
+            // Reset the motor limits
+            SetMotorLimits(MotorPort.Both, 0, 0);
+            // Turn off the servos
+            SetServo(ServoPort.Both, 0);
+            // Turn off the LEDs
+            SetLed((byte)GoPiGo3Led.LedEyeLeft + (byte)GoPiGo3Led.LedEyeRight + (byte)GoPiGo3Led.LedBlinkerLeft + (byte)GoPiGo3Led.LedBlinkerRight, Color.Black);
+        }
+
+        #endregion
+
+        #region SPI transfer
+
+        /// <summary>
+        /// Conduct a SPI transaction
+        /// </summary>
+        /// <param name="dataOut">A byte array to send.The length of the array will determine how many bytes are transferred.</param>
+        /// <returns>Returns an array of the bytes read.</returns>
+        public byte[] SpiTransferArray(byte[] dataOut)
+        {
+            byte[] result = new byte[dataOut.Length];
+            _spiDevice.TransferFullDuplex(dataOut, result);
+            return result;
+        }
+
+        /// <summary>
+        /// Read a 32 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <returns>Returns the value read</returns>
+        public int SpiRead32(SpiMessageType MessageType)
```

Comment:
Same comment about this where it shouldn't be public surface area. Same comment applies to all of your public methods, so I won't be adding more comments about it but you should do a global check to make sure you only expose high level APIs.

Reply:
I though we had something for that but couldn't find it :smile:

Reply:
I'll keep those ones as if you have your own sensors, you willhave to call those low level functions. But I'll private the SpiTransfer one